File: go/internal/tui/tui.go
package tui

import (
	"context"
	"fmt"
	"strings"

	tea "github.com/charmbracelet/bubbletea"
	"github.com/charmbracelet/lipgloss"

	"github.com/foolish-dev/Teleia/go/internal/agent"
)

const hints = "enter send · ↑↓ scroll · /help cmds · ctrl-c quit"

var (
	userStyle      = lipgloss.NewStyle().Foreground(lipgloss.Color("#7dcfff")).Bold(true)
	assistantStyle = lipgloss.NewStyle().Foreground(lipgloss.Color("#bb9af7")).Bold(true)
	toolStyle      = lipgloss.NewStyle().Foreground(lipgloss.Color("#e0af68"))
	dimStyle       = lipgloss.NewStyle().Foreground(lipgloss.Color("#565f89"))
	errorStyle     = lipgloss.NewStyle().Foreground(lipgloss.Color("#f7768e"))
	infoStyle      = lipgloss.NewStyle().Foreground(lipgloss.Color("#7aa2f7"))
	promptStyle    = lipgloss.NewStyle().Foreground(lipgloss.Color("#7dcfff"))
	borderStyle    = lipgloss.NewStyle().Border(lipgloss.RoundedBorder()).Padding(0, 1)
)

type entry struct {
	kind     string // user, assistant, tool, error, info
	text     string
	complete bool
	tool     toolDetails
}

type toolDetails struct {
	name      string
	arguments string
	output    string
}

type eventMsg struct{ ev agent.TurnEvent }

type errMsg struct{ err error }

type model struct {
	agent   *agent.Agent
	width   int
	height  int
	input   string
	history []entry
	status  string
	working bool
	scroll  int
	stream  <-chan agent.TurnEvent
}

func Run(a *agent.Agent) error {
	m := model{
		agent:  a,
		status: fmt.Sprintf("session %s · ready", shortID(a.SessionID())),
	}
	_, err := tea.NewProgram(m, tea.WithAltScreen()).Run()
	return err
}

func shortID(s string) string {
	if len(s) <= 12 {
		return s
	}
	return s[:12]
}

func (m model) Init() tea.Cmd { return nil }

func (m model) Update(msg tea.Msg) (tea.Model, tea.Cmd) {
	switch msg := msg.(type) {
	case tea.WindowSizeMsg:
		m.width, m.height = msg.Width, msg.Height

	case eventMsg:
		m.applyEvent(msg.ev)
		if _, isEnd := msg.ev.(agent.TurnEnd); isEnd {
			m.working = false
			m.stream = nil
			m.status = fmt.Sprintf("session %s · ready", shortID(m.agent.SessionID()))
			return m, nil
		}
		return m, recvEvent(m.stream)

	case errMsg:
		m.history = append(m.history, entry{kind: "error", text: msg.err.Error()})
		m.working = false
		m.stream = nil
		m.status = "error · ready"
		return m, nil

	case tea.KeyMsg:
		if msg.Type == tea.KeyCtrlC {
			return m, tea.Quit
		}
		if m.working {
			return m, nil
		}
		switch msg.Type {
		case tea.KeyUp:
			m.scroll++
		case tea.KeyDown:
			if m.scroll > 0 {
				m.scroll--
			}
		case tea.KeyPgUp:
			m.scroll += 5
		case tea.KeyPgDown:
			if m.scroll > 5 {
				m.scroll -= 5
			} else {
				m.scroll = 0
			}
		case tea.KeyEnter:
			raw := strings.TrimSpace(m.input)
			m.input = ""
			if raw == "" {
				return m, nil
			}
			if strings.HasPrefix(raw, "/") {
				m.handleSlash(raw[1:])
				return m, nil
			}
			m.history = append(m.history, entry{kind: "user", text: raw})
			m.scroll = 0
			m.working = true
			m.status = "thinking…"
			stream, err := m.agent.Turn(context.Background(), raw)
			if err != nil {
				return m, func() tea.Msg { return errMsg{err: err} }
			}
			m.stream = stream
			return m, recvEvent(stream)
		case tea.KeyBackspace:
			if len(m.input) > 0 {
				m.input = m.input[:len(m.input)-1]
			}
		case tea.KeySpace:
			m.input += " "
		case tea.KeyRunes:
			m.input += string(msg.Runes)
		}
	}
	return m, nil
}

func recvEvent(stream <-chan agent.TurnEvent) tea.Cmd {
	return func() tea.Msg {
		ev, ok := <-stream
		if !ok {
			return eventMsg{ev: agent.TurnEnd{}}
		}
		return eventMsg{ev: ev}
	}
}

func (m *model) applyEvent(ev agent.TurnEvent) {
	switch v := ev.(type) {
	case agent.AssistantStart:
		m.history = append(m.history, entry{kind: "assistant", text: "", complete: false})
	case agent.AssistantDelta:
		if n := len(m.history); n > 0 && m.history[n-1].kind == "assistant" && !m.history[n-1].complete {
			m.history[n-1].text += v.Text
		} else {
			m.history = append(m.history, entry{kind: "assistant", text: v.Text})
		}
		m.scroll = 0
	case agent.AssistantEnd:
		if n := len(m.history); n > 0 && m.history[n-1].kind == "assistant" {
			m.history[n-1].complete = true
			if m.history[n-1].text == "" {
				m.history = m.history[:n-1]
			}
		}
	case agent.ToolStart:
		m.history = append(m.history, entry{
			kind: "tool", tool: toolDetails{name: v.Name, arguments: v.Arguments},
		})
		m.scroll = 0
	case agent.ToolEnd:
		if n := len(m.history); n > 0 && m.history[n-1].kind == "tool" {
			m.history[n-1].tool.output = v.Output
			m.history[n-1].complete = true
			m.scroll = 0
		}
	}
}

func (m *model) handleSlash(cmd string) {
	name, _, _ := strings.Cut(cmd, " ")
	name = strings.TrimSpace(name)
	arg := strings.TrimSpace(strings.TrimPrefix(cmd, name))

	push := func(kind, text string) {
		m.history = append(m.history, entry{kind: kind, text: text})
		m.scroll = 0
	}

	switch name {
	case "reset":
		if err := m.agent.Reset(); err != nil {
			push("error", "reset: "+err.Error())
			return
		}
		m.history = nil
		push("info", "started new session "+shortID(m.agent.SessionID()))
	case "save":
		if arg == "" {
			push("error", "usage: /save NAME")
			return
		}
		if err := m.agent.SaveAlias(arg); err != nil {
			push("error", "save: "+err.Error())
			return
		}
		push("info", "saved current session as '"+arg+"'")
	case "load":
		if arg == "" {
			push("error", "usage: /load NAME")
			return
		}
		id, err := m.agent.LoadAlias(arg)
		if err != nil {
			push("error", "load: "+err.Error())
			return
		}
		m.history = nil
		push("info", "loaded '"+arg+"' → session "+shortID(id))
	case "help", "?":
		push("info", "commands: /reset · /save NAME · /load NAME · /help")
	default:
		push("error", "unknown command: /"+name)
	}
}

func (m model) View() string {
	if m.width == 0 {
		return ""
	}
	logH := m.height - 4
	if logH < 3 {
		logH = 3
	}

	var lines []string
	for _, e := range m.history {
		lines = append(lines, renderEntry(e)...)
	}

	maxOffset := len(lines) - logH
	if maxOffset < 0 {
		maxOffset = 0
	}
	start := maxOffset - m.scroll
	if start < 0 {
		start = 0
	}
	end := start + logH
	if end > len(lines) {
		end = len(lines)
	}
	log := strings.Join(lines[start:end], "\n")
	logBox := borderStyle.Width(m.width - 2).Height(logH).Render(log)

	inputLine := promptStyle.Render("> ") + m.input
	inputBox := borderStyle.Width(m.width - 2).Render(inputLine)

	status := dimStyle.Render(m.status + "   " + hints)
	return lipgloss.JoinVertical(lipgloss.Left, logBox, inputBox, status)
}

func renderEntry(e entry) []string {
	var out []string
	switch e.kind {
	case "user":
		out = append(out, userStyle.Render("you"))
		out = append(out, e.text, "")
	case "assistant":
		header := "teleia"
		if !e.complete {
			header = "teleia ▌"
		}
		out = append(out, assistantStyle.Render(header))
		out = append(out, e.text, "")
	case "tool":
		marker := "⚙"
		if !e.complete {
			marker = "⚙ …"
		}
		out = append(out, toolStyle.Render(fmt.Sprintf("%s %s(%s)", marker, e.tool.name, e.tool.arguments)))
		count := 0
		for _, l := range strings.Split(e.tool.output, "\n") {
			if count >= 20 {
				break
			}
			out = append(out, dimStyle.Render("  "+l))
			count++
		}
		out = append(out, "")
	case "error":
		out = append(out, errorStyle.Render("error: "+e.text), "")
	case "info":
		out = append(out, infoStyle.Render("· "+e.text), "")
	}
	return out
}

File: rust/crates/teleia-cli/src/tui.rs
use anyhow::Result;
use crossterm::{
    event::{
        self, DisableMouseCapture, EnableMouseCapture, Event, KeyCode, KeyEventKind, KeyModifiers,
    },
    execute,
    terminal::{disable_raw_mode, enable_raw_mode, EnterAlternateScreen, LeaveAlternateScreen},
};
use futures_util::{pin_mut, StreamExt};
use ratatui::{
    backend::CrosstermBackend,
    layout::{Constraint, Direction, Layout},
    style::{Color, Modifier, Style},
    text::{Line, Span},
    widgets::{Block, Borders, Paragraph, Wrap},
    Terminal,
};
use std::{io, time::Duration};
use teleia_agent::{Agent, TurnEvent};

const HINTS: &str = "enter send · ↑↓ scroll · /help cmds · ctrl-c quit";

enum Entry {
    User(String),
    Assistant {
        text: String,
        complete: bool,
    },
    Tool {
        name: String,
        args: String,
        output: String,
        complete: bool,
    },
    Error(String),
    Info(String),
}

struct State {
    input: String,
    history: Vec<Entry>,
    status: String,
    scroll: u16, // offset from auto-scroll bottom; 0 = follow
    working: bool,
}

impl State {
    fn new(session_id: &str) -> Self {
        Self {
            input: String::new(),
            history: Vec::new(),
            status: format!(
                "session {} · ready",
                &session_id[..session_id.len().min(12)]
            ),
            scroll: 0,
            working: false,
        }
    }

    fn push(&mut self, e: Entry) {
        self.history.push(e);
        self.scroll = 0; // auto-scroll on new content
    }

    fn apply(&mut self, evt: TurnEvent) {
        match evt {
            TurnEvent::AssistantStart => {
                self.push(Entry::Assistant {
                    text: String::new(),
                    complete: false,
                });
            }
            TurnEvent::AssistantDelta(text) => {
                if let Some(Entry::Assistant {
                    text: t,
                    complete: false,
                }) = self.history.last_mut()
                {
                    t.push_str(&text);
                    self.scroll = 0;
                } else {
                    self.push(Entry::Assistant {
                        text,
                        complete: false,
                    });
                }
            }
            TurnEvent::AssistantEnd => {
                if let Some(Entry::Assistant { complete, text }) = self.history.last_mut() {
                    *complete = true;
                    if text.is_empty() {
                        self.history.pop();
                    }
                }
            }
            TurnEvent::ToolStart { name, arguments } => {
                self.push(Entry::Tool {
                    name,
                    args: arguments,
                    output: String::new(),
                    complete: false,
                });
            }
            TurnEvent::ToolEnd { output, .. } => {
                if let Some(Entry::Tool {
                    output: o,
                    complete,
                    ..
                }) = self.history.last_mut()
                {
                    *o = output;
                    *complete = true;
                    self.scroll = 0;
                }
            }
            TurnEvent::TurnEnd => {}
        }
    }
}

pub async fn run(mut agent: Agent) -> Result<()> {
    enable_raw_mode()?;
    let mut stdout = io::stdout();
    execute!(stdout, EnterAlternateScreen, EnableMouseCapture)?;
    let backend = CrosstermBackend::new(stdout);
    let mut terminal = Terminal::new(backend)?;

    let mut state = State::new(agent.session_id());
    let result = event_loop(&mut terminal, &mut state, &mut agent).await;

    disable_raw_mode()?;
    execute!(
        terminal.backend_mut(),
        LeaveAlternateScreen,
        DisableMouseCapture
    )?;
    terminal.show_cursor()?;
    result
}

async fn event_loop<B: ratatui::backend::Backend>(
    terminal: &mut Terminal<B>,
    state: &mut State,
    agent: &mut Agent,
) -> Result<()> {
    loop {
        terminal.draw(|f| draw(f, state))?;

        if !event::poll(Duration::from_millis(50))? {
            continue;
        }
        let Event::Key(key) = event::read()? else {
            continue;
        };
        if key.kind == KeyEventKind::Release {
            continue;
        }

        if key.modifiers.contains(KeyModifiers::CONTROL) && matches!(key.code, KeyCode::Char('c')) {
            return Ok(());
        }

        match key.code {
            KeyCode::Up | KeyCode::PageUp => {
                state.scroll =
                    state
                        .scroll
                        .saturating_add(if key.code == KeyCode::PageUp { 5 } else { 1 });
            }
            KeyCode::Down | KeyCode::PageDown => {
                state.scroll = state
                    .scroll
                    .saturating_sub(if key.code == KeyCode::PageDown { 5 } else { 1 });
            }
            KeyCode::Char(c) => state.input.push(c),
            KeyCode::Backspace => {
                state.input.pop();
            }
            KeyCode::Enter => {
                let raw = std::mem::take(&mut state.input);
                let trimmed = raw.trim();
                if trimmed.is_empty() {
                    continue;
                }
                if let Some(cmd) = trimmed.strip_prefix('/') {
                    handle_slash(state, agent, cmd);
                    continue;
                }
                state.push(Entry::User(trimmed.to_string()));
                state.status = "thinking…".into();
                state.working = true;
                run_turn(terminal, state, agent, trimmed.to_string()).await;
                state.working = false;
                state.status = format!(
                    "session {} · ready",
                    &agent.session_id()[..agent.session_id().len().min(12)]
                );
            }
            _ => {}
        }
    }
}

async fn run_turn<B: ratatui::backend::Backend>(
    terminal: &mut Terminal<B>,
    state: &mut State,
    agent: &mut Agent,
    input: String,
) {
    let stream = agent.turn(input);
    pin_mut!(stream);
    loop {
        if let Err(e) = terminal.draw(|f| draw(f, state)) {
            state.push(Entry::Error(format!("draw: {e:#}")));
            return;
        }

        tokio::select! {
            evt = stream.next() => {
                match evt {
                    Some(Ok(e)) => {
                        let is_end = matches!(e, TurnEvent::TurnEnd);
                        state.apply(e);
                        if is_end { return; }
                    }
                    Some(Err(e)) => {
                        state.push(Entry::Error(e.to_string()));
                        return;
                    }
                    None => return,
                }
            }
            _ = tokio::time::sleep(Duration::from_millis(33)) => {
                // periodic redraw so deltas land smoothly
            }
        }
    }
}

fn handle_slash(state: &mut State, agent: &mut Agent, cmd: &str) {
    let mut parts = cmd.splitn(2, char::is_whitespace);
    let name = parts.next().unwrap_or("").trim();
    let arg = parts.next().unwrap_or("").trim();
    match name {
        "reset" => match agent.reset() {
            Ok(()) => {
                state.history.clear();
                state.push(Entry::Info(format!(
                    "started new session {}",
                    &agent.session_id()[..agent.session_id().len().min(12)]
                )));
            }
            Err(e) => state.push(Entry::Error(format!("reset: {e}"))),
        },
        "save" => {
            if arg.is_empty() {
                state.push(Entry::Error("usage: /save NAME".into()));
                return;
            }
            match agent.save_alias(arg) {
                Ok(()) => state.push(Entry::Info(format!("saved current session as '{arg}'"))),
                Err(e) => state.push(Entry::Error(format!("save: {e}"))),
            }
        }
        "load" => {
            if arg.is_empty() {
                state.push(Entry::Error("usage: /load NAME".into()));
                return;
            }
            match agent.load_alias(arg) {
                Ok(id) => {
                    state.history.clear();
                    state.push(Entry::Info(format!(
                        "loaded '{arg}' → session {}",
                        &id[..id.len().min(12)]
                    )));
                }
                Err(e) => state.push(Entry::Error(format!("load: {e}"))),
            }
        }
        "help" | "?" => {
            state.push(Entry::Info(
                "commands: /reset · /save NAME · /load NAME · /help".into(),
            ));
        }
        other => {
            state.push(Entry::Error(format!("unknown command: /{other}")));
        }
    }
}

fn draw(f: &mut ratatui::Frame, state: &State) {
    let chunks = Layout::default()
        .direction(Direction::Vertical)
        .constraints([
            Constraint::Min(3),
            Constraint::Length(3),
            Constraint::Length(1),
        ])
        .split(f.area());

    let lines: Vec<Line> = state.history.iter().flat_map(render_entry).collect();
    let visible = chunks[0].height.saturating_sub(2) as usize;
    let total = lines.len();
    let max_offset = total.saturating_sub(visible) as u16;
    let offset = max_offset.saturating_sub(state.scroll);

    let log = Paragraph::new(lines)
        .block(Block::default().borders(Borders::ALL).title(" teleia "))
        .wrap(Wrap { trim: false })
        .scroll((offset, 0));
    f.render_widget(log, chunks[0]);

    let prompt_style = if state.working {
        Style::default().fg(Color::DarkGray)
    } else {
        Style::default().fg(Color::White)
    };
    let input_widget = Paragraph::new(Line::from(vec![
        Span::styled("> ", Style::default().fg(Color::Cyan)),
        Span::styled(&state.input, prompt_style),
    ]))
    .block(Block::default().borders(Borders::ALL));
    f.render_widget(input_widget, chunks[1]);

    let status_line = Line::from(vec![
        Span::styled(
            &state.status,
            Style::default()
                .fg(Color::DarkGray)
                .add_modifier(Modifier::ITALIC),
        ),
        Span::raw("   "),
        Span::styled(HINTS, Style::default().fg(Color::DarkGray)),
    ]);
    f.render_widget(Paragraph::new(status_line), chunks[2]);
}

fn render_entry(entry: &Entry) -> Vec<Line<'static>> {
    let mut out = Vec::new();
    match entry {
        Entry::User(text) => {
            out.push(Line::from(Span::styled(
                "you",
                Style::default()
                    .fg(Color::Cyan)
                    .add_modifier(Modifier::BOLD),
            )));
            for line in text.lines() {
                out.push(Line::from(line.to_string()));
            }
            out.push(Line::from(""));
        }
        Entry::Assistant { text, complete } => {
            out.push(Line::from(Span::styled(
                if *complete { "teleia" } else { "teleia ▌" },
                Style::default()
                    .fg(Color::Magenta)
                    .add_modifier(Modifier::BOLD),
            )));
            for line in text.lines() {
                out.push(Line::from(line.to_string()));
            }
            out.push(Line::from(""));
        }
        Entry::Tool {
            name,
            args,
            output,
            complete,
        } => {
            let marker = if *complete { "⚙" } else { "⚙ …" };
            out.push(Line::from(Span::styled(
                format!("{marker} {name}({args})"),
                Style::default().fg(Color::Yellow),
            )));
            for line in output.lines().take(20) {
                out.push(Line::from(Span::styled(
                    format!("  {line}"),
                    Style::default().fg(Color::DarkGray),
                )));
            }
            out.push(Line::from(""));
        }
        Entry::Error(text) => {
            out.push(Line::from(Span::styled(
                format!("error: {text}"),
                Style::default().fg(Color::Red),
            )));
            out.push(Line::from(""));
        }
        Entry::Info(text) => {
            out.push(Line::from(Span::styled(
                format!("· {text}"),
                Style::default().fg(Color::Blue),
            )));
            out.push(Line::from(""));
        }
    }
    out
}
